### PR TITLE
feat(audio-only): be able to lock a browser into capturing audio only

### PR DESCRIPTION
**Pull request by virtuacoplenny**
relies on https://github.com/jitsi/lib-jitsi-meet/pull/629

### CODE REVIEW

File: lang/main.json [hunk 1]
```diff
@@ -452,6 +452,8 @@
         "unlocked": "This call is unlocked. Any new caller with the link may join the call."
     },
     "videoStatus": {
+        "audioOnlyLocked": "Locked in audio only mode",
+        "audioOnlySupported": "The current browser supports only audio.",
```
Comment (bbaldino) [inline lang/main.json]: `The current browser only supports audio` ?

File: react/features/base/conference/actions.js [hunk 2]
```diff
@@ -365,19 +366,45 @@ export function p2pStatusChanged(p2p: boolean) {
 }
 
 /**
+ * Sets a flag for the current JitsiConference to never allow disabling
+ * of audio only mode.
+ *
+ * @returns {Function}
+ */
+export function enableAudioOnlyLock() {
```
Comment (bbaldino) [inline react/features/base/conference/actions.js]: as we talked about offline, maybe we want this named/done a bit differently.  maybe have `RTCBrowserType.supportsVideo()` and key off of that?

File: react/features/base/conference/actions.js [hunk 3]
```diff
@@ -365,19 +366,45 @@ export function p2pStatusChanged(p2p: boolean) {
...
  */
-export function setAudioOnly(audioOnly: boolean) {
-    return {
-        type: SET_AUDIO_ONLY,
-        audioOnly
+export function setAudioOnly(isAudioOnly: boolean) {
+    return (dispatch: Dispatch<*>, getState: Function) => {
+        const {
+            audioOnly,
+            audioOnlyLockEnabled
```
Comment (bbaldino) [inline react/features/base/conference/actions.js]: same as note above...maybe we can simplify the variables/states needed here

**Comment by virtuacoplenny** [comment]
@saghul, could I get your thoughts on this implementation? Safari + adapter will be audio only. To enforce such, there are changes in lib-jitsi-meet to strip out audio devices from getUserMedia calls. On the jitsi-meet side, there has to be additional handling to to keep the UI consistent with the forced audio only state. What I did was take the existing audioOnly state in redux and have it be undefined (initial state), false, or a string that differentiates between enabled+locked and enabled+unlocked. I could force it to be undefined whenever audioOnly is disabled.

File: react/features/base/media/middleware.js [hunk 4]
```diff
@@ -113,20 +113,21 @@ function _setRoom({ dispatch, getState }, next, action) {
...
     // config.startWithVideoMuted.
     if (room) {
-        const supportsOnlyAudio = JitsiMeetJS.mediaDevices.supportsOnlyAudio();
+        const supportsVideo = JitsiMeetJS.mediaDevices.supportsVideo();
 
-        let audioOnly = urlParams && urlParams['config.startAudioOnly'];
+        let audioOnly = !supportsVideo
+            || (urlParams && urlParams['config.startAudioOnly']);
 
         typeof audioOnly === 'undefined' && (audioOnly = config.startAudioOnly);
```
Comment (bbaldino) [inline react/features/base/media/middleware.js]: maybe we can combine this with the assignment on 118

File: react/features/base/media/middleware.js [hunk 5]
```diff
@@ -113,20 +113,21 @@ function _setRoom({ dispatch, getState }, next, action) {
...
             `startaudioonly.${audioOnly ? 'enabled' : 'disabled'}`);
         logger.log(`Start audio only set to ${audioOnly.toString()}`);
 
-        if (supportsOnlyAudio) {
-            dispatch(setAudioOnlyWithLock());
-        } else {
+        if (supportsVideo) {
             dispatch(setAudioOnly(audioOnly));
+        } else {
+            dispatch(setAudioOnlyWithLock());
```
Comment (bbaldino) [inline react/features/base/media/middleware.js]: what about `setPermanentAudioOnly` ?

**Comment by saghul** [comment]
@virtuacoplenny I'll take a look, thanks for the heads up!

File: react/features/base/conference/actions.js [hunk 6]
```diff
@@ -378,30 +378,31 @@ export function setAudioOnly(isAudioOnly: boolean) {
     return (dispatch: Dispatch<*>, getState: Function) => {
         const { audioOnly } = getState()['features/base/conference'];
 
-        const isLocked = audioOnly === AUDIO_ONLY_WITH_LOCK;
+        const isPermenantAudioOnly
+            = audioOnly === AUDIO_ONLY_ENABLED_PERMANENTLY;
 
-        if (!isLocked || (isLocked && !audioOnly)) {
+        if (!isPermenantAudioOnly || (isPermenantAudioOnly && !audioOnly)) {
```
Comment (bbaldino) [inline react/features/base/conference/actions.js]: this block is confusing to read

File: react/features/base/conference/actions.js [hunk 7]
```diff
@@ -378,30 +378,31 @@ export function setAudioOnly(isAudioOnly: boolean) {
...
             dispatch({
                 type: SET_AUDIO_ONLY,
-                audioOnly: isAudioOnly ? AUDIO_ONLY_WITHOUT_LOCK : isAudioOnly
+                audioOnly: isAudioOnly ? AUDIO_ONLY_ENABLED : isAudioOnly
```
Comment (bbaldino) [inline react/features/base/conference/actions.js]: confused by the mixture of using a constant and the `isAudioOnly` value

**Comment by bbaldino** [comment]
i think the use of 'permanent(ly)' is definitely clearer, but still some blocks of code in there that end up being hard to parse because of all the similar values i think

**Comment by virtuacoplenny** [comment]
Do you think going back to having a separate value for permanentlyAudioOnly would help clarify some of the confusion?

**Comment by virtuacoplenny** [comment]
I might have to rebase and force push to resolve the failing build... I'll do that once discussion on implementation has concluded.

File: react/features/base/conference/actions.js [hunk 8]
```diff
@@ -370,21 +369,19 @@ export function p2pStatusChanged(p2p: boolean) {
...
  * @returns {Function}
  */
-export function setAudioOnly(isAudioOnly: boolean) {
+export function setAudioOnly(audioOnly: boolean) {
     return (dispatch: Dispatch<*>, getState: Function) => {
-        const { audioOnly } = getState()['features/base/conference'];
-
-        const isPermenantAudioOnly
-            = audioOnly === AUDIO_ONLY_ENABLED_PERMANENTLY;
+        const currentlyAudioOnlyState
```
Comment (bbaldino) [inline react/features/base/conference/actions.js]: should this be `currentAudioOnlyState` ?

File: react/features/base/conference/actions.js [hunk 9]
```diff
@@ -399,10 +396,11 @@ export function setAudioOnly(isAudioOnly: boolean) {
  *     audioOnly: string
  * }}
  */
-export function setPermanentAudioOnly() {
+export function setPermanentlyAudioOnly() {
```
Comment (bbaldino) [inline react/features/base/conference/actions.js]: i liked the old name better i think.  but fine either way.

File: react/features/base/conference/constants.js [hunk 10]
```diff
@@ -1,16 +1,16 @@
 /**
- * The audio only state where audio only mode is enabled and cannot be disabled.
+ * The audio only state where audio only mode is enabled and can be disabled.
  *
  * @type {string}
  */
-export const AUDIO_ONLY_ENABLED_PERMANENTLY = 'audio-only-enabled-with-lock';
+export const AUDIO_ONLY_ENABLED = 'audio-only-enabled-without-lock';
```
Comment (bbaldino) [inline react/features/base/conference/constants.js]: should we change the name of the field to match the new variable names? ('permanent' vs 'lock')

File: react/features/base/conference/reducer.js [hunk 11]
```diff
@@ -303,7 +307,16 @@ function _p2pStatusChanged(state, action) {
...
  */
 function _setAudioOnly(state, action) {
-    return set(state, 'audioOnly', action.audioOnly);
+    let newAudioOnlyState;
+
+    if (action.audioOnly) {
+        newAudioOnlyState = action.permanent
+            ? AUDIO_ONLY_ENABLED_PERMANENTLY : AUDIO_ONLY_ENABLED;
+    } else {
+        newAudioOnlyState = undefined;
```
Comment (bbaldino) [inline react/features/base/conference/reducer.js]: can this else just be removed completely? it'll be undefined by default, right?

**Comment by hristoterezov** [comment]
@virtuacoplenny Could you please rebase with the latest master in order to get the package-lock file?

**Comment by virtuacoplenny** [comment]
Rebased, squashed, and force pushed.

**Comment by virtuacoplenny** [comment]
I tried out a new implementation. The original idea was to prevent any mechanism from allowing the user to get out of audio only mode, so the store and actions were changed, but this new batch of changes disables the slider based on video support.

Safari webrtc does support H264, however browsers by default strip H264 (config.p2p.disableH264) and the bridge does not send H264.

**Comment by gpolitis** [comment]
I'm not sure how far we're into the existing approach and what would it take to change it, but I'm feel more confortable with what @saghul is suggesting, than to detect safari and forcing no video.

**Comment by virtuacoplenny** [comment]
@gpolitis, there are a few concerns to consider when asking for video to work on safari. First is that safari only has H264 support, no VP8/VP9. This is not a problem at all if everyone offered H264. The question becomes how would that be implemented? So you may capture video initially, before joining a conference. Then someone else joins and you would maybe look in the offer and toggle video support based on what's available (or assume based on config). So in a P2P situation, safari might see H264 is supported and so sends/receives video. Then when another user joins, the bridge offers without H264 and then you modify the UX to explain why video no longer works. This also still encounters the corner case where webkit errors when there are no supported codecs in the offer, so you'd still want to inject H264 to prevent the error (or maybe you figured out a different workaround), but this injection logic would be conditionally avoided if H264 is present. I'm assuming to avoid having conditional toggling of video support in some situations but not others, the requirement was set to have safari support audio only. If you feel strongly about video support, then please do bring it up because I'm open to it but I'm uncertain the work involved is worthwhile.

**Comment by saghul** [comment]
> I tried out a new implementation. The original idea was to prevent any mechanism from allowing the user to get out of audio only mode, so the store and actions were changed, but this new batch of changes disables the slider based on video support.

❤️ ❤️ ❤️

File: lang/main.json [hunk 12]
```diff
@@ -466,6 +466,8 @@
         "unlocked": "This call is unlocked. Any new caller with the link may join the call."
     },
     "videoStatus": {
+        "audioOnlyLocked": "Locked in audio only mode",
+        "audioOnlySupported": "The current browser only supports audio.",
```
Comment (saghul) [inline lang/main.json]: Maybe word this differently so it's clearer that **we** don't support video in the current browser.

**Comment by virtuacoplenny** [comment]
I tried fixing the string. Maybe it's better. I also rebased to pick up the updated package-lock which includes the lib-jitsi-meet side of these changes.

**Comment by virtuacoplenny** [comment]
Currently waiting on final strings.
